### ISSUE TITLE
Joint Dimension and Unit Mismatch during Inference with LeIsaac and Pi0

### ISSUE THREAD

**Issue by Nadeem752** [LightwheelAI/leisaac#136]
https://github.com/user-attachments/assets/394efc43-ef85-4c37-a10b-9605f1fd1ba9

Model: Pi0 (Finetuned on 2 cameras: front_camera and wrist_camera)

Robot: SO-100 (6 DoF)

Remote Policy Server (websocket_policy_server.py)

I am experiencing a coordination failure when running a finetuned Pi0 model via the LeIsaac inference script. While the model was trained on a simulation dataset and the inference server is successfully receiving observations, the robot exhibits two primary failure modes:

The arm reaches toward the target (the orange) but hovers near it without making contact or grasping. The gripper remains permanently open, and the arm occasionally exhibits high-frequency "jitter" or thrashing.

After debugging the obs_dict and action_tensor on both the client and server sides, I identified three critical mismatches:

The model appears to be outputting values that the simulation interprets as being in the wrong unit space (potentially a Radians vs. Degrees mismatch or Absolute vs. Velocity mismatch).

Example log output: RAW ACTION SENT TO ROBOT: tensor([[11.1838, -4.9166, 18.0119, 18.8148, -1.3246, 6.3647]]). These values exceed typical radian limits for the SO-100 joints in IsaacLab, causing the physics engine to saturate the motors.

**Comment by Nadeem752** [comment]
@EverNorif  Hey could you please assist ? I am really stuck here :( . Ihave checked my service_policy_clients.py file and policy _inference.py file . I also checked that my pi0 was trained on deltas . I check the images that are being passed to the model but still no luck :(

**Comment by EverNorif** [comment]
Hi @Nadeem752 Could you provide more information so that we can debug it? Including the version of the pi0 code you are using, and the version of the leisaac code. If possible, could you share the fine-tuning dataset you are currently using on Hugging Face?

In addition, here is some information that might be useful for you. In our previous tests, the dataset we used can be found [here](https://huggingface.co/datasets/LightwheelAI/leisaac-pick-orange), and the [code](https://github.com/EverNorif/openpi/tree/lerobot-v0.3.3) we used.

**Comment by Nadeem752** [comment]
Hi, please find the requested details below -
open pi -Openpi$ git status
HEAD detached at 5bff19b
lerobot - 0.3.3 for leisaac(venv) and 0.1.0 for openpi (venv)
git clone https://huggingface.co/datasets/Nadeem-Shoukath115113/so101_tm7

**Comment by Nadeem752** [comment]
my dataset on Groot 1.5 works 

https://github.com/user-attachments/assets/f700e690-13fa-4fa0-9dad-4196f30d2384

**Comment by Nadeem752** [comment]
and pi0 on Pick-orange works somewhat :-

https://github.com/user-attachments/assets/91aae4c2-92b4-4056-82d2-fa2e11959b83

**Comment by CartmanRolex** [comment]
I had the same issues but with smolvla. Since I am doing inference with a custom python script, I fixed it by converting the output from degrees to radians, and converting the observation joint space into degrees from radians. I still have an issue with the camera view(the observation gives a view that has a strange rendering which is quite different from the dataset), which explains why the model struggles to find the orange.

**Comment by EverNorif** [comment]
@Nadeem752 Hi, I checked the information you provided.

1. The data range in the lerobot dataset is correct and corresponds to `angles`.
2. The version of openpi is basically the same as the one we are using.
3. Regarding the unit space error issue you discovered, openpi is trained using data where the units are `angles`. During inference, the joint positions in the obs_dict sent by leisaac to the model are converted to angles through [convert_leisaac_action_to_lerobot](https://github.com/LightwheelAI/leisaac/blob/main/source/leisaac/leisaac/policy/service_policy_clients.py#L372). Additionally, the actions obtained from the model are converted to `radians` through [convert_lerobot_action_to_leisaac](https://github.com/LightwheelAI/leisaac/blob/main/source/leisaac/leisaac/policy/service_policy_clients.py#L395) before being passed to isaaclab’s `env.step`. I believe the inference logic is correct, but you can also double-check it.

I suspect the issue might lie in the data and the training approach. Perhaps these datasets are not sufficient to successfully fine-tune pi0 (even though they worked on gr00t). You could try collecting more data, or compare the performance of full fine-tuning and LoRA fine-tuning.

**Comment by LeeZheXian** [comment]
Hello @Nadeem752 can u please share with me your gr00t's N1.5config? mine is behaving like your pi0 model, it will be a great help! Thank you!

![Image](https://github.com/user-attachments/assets/ed2b1298-a6d7-4f25-8d94-eb8390f29b9c)

**Comment by Nadeem752** [comment]
Hello @LeeZheXian  , which file do you need ? I moved on to a different project so I dont remember which specific file you are asking right haha sorry . I ran these commands and it worked right out of the box I think -`PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True \
python scripts/gr00t_finetune.py \
   --dataset-path /home/nadeem/.cache/huggingface/lerobot/EverNorif/so101_tm7 \
   --num-gpus 1 \
   --output-dir ./so101-checkpoints \
   --max-steps 30000 \
   --data-config so100_dualcam \
   --video-backend torchvision_av \
   --batch-size 4 \
   --gradient-accumulation-steps 16 \
   --dataloader-num-workers 16 \
   --lora-rank 32
   
   
python scripts/inference_service.py \
    --server \
    --model_path /home/nadeem/IsaacLab/leisaac/Isaac-GR00T/so101-checkpoints\
    --embodiment_tag new_embodiment \
    --data_config so100_dualcam 

python scripts/evaluation/policy_inference.py \
    --task=LeIsaac-SO101-tm7Task-v0 \
    --eval_rounds=10 \
    --policy_type=gr00tn1.5 \
    --policy_host=localhost \
    --policy_port=5555 \
    --policy_timeout_ms=5000 \
    --policy_action_horizon=16 \
    --policy_language_instruction="pick up the orange and place it into the black bowl." \
    --device=cuda \
    --enable_cameras`
